File: zig/src/p_0004.zig
// A palindromic number reads the same both ways. The largest palindrome made
// from the product of two 2-digit numbers is 9009 = 91 × 99.
//
// Find the largest palindrome made from the product of two 3-digit numbers.

const std = @import("std");
const fmt = std.fmt;

const IntPair = struct { first: usize, second: usize };

fn is_number_palindrome(number: usize) bool {
    var buf: [128]u8 = undefined;

    const value = fmt.bufPrint(&buf, "{}", .{number}) catch unreachable;

    const len: usize = value.len - 1;
    for (value, 0..) |c, i| {
        if (c != value[len - i]) return false;
    }

    return true;
}

fn euler_0004(digits: usize) IntPair {
    var limit: usize = 1;
    for (0..digits) |_| {
        limit *= 10;
    }

    var result: IntPair = .{ .first = 0, .second = 0 };
    var largest_palindrome: usize = 0;
    var value: usize = 0;
    for ((limit / 10)..limit) |i| {
        for (i..limit) |j| {
            value = i * j;
            if (is_number_palindrome(value) and value > largest_palindrome) {
                largest_palindrome = value;
                result = .{ .first = i, .second = j };
            }
        }
    }

    return result;
}

pub fn main() anyerror!void {
    var out = std.fs.File.stdout().writerStreaming(&.{});
    const factors = euler_0004(3);
    const i = factors.first;
    const j = factors.second;
    try out.interface.print("Problem 0004: The largest palindrome made from the product of two 3-digit numbers is: {d} x {d} = {d}\n", .{ i, j, i * j });
}

test "the largest palindrome made from product of 2-digit numbers is 9009" {
    const factors = euler_0004(2);
    const i = factors.first;
    const j = factors.second;
    try std.testing.expect((i * j) == 9009);
}

File: cpp/src/p_0004.cc
// A palindromic number reads the same both ways. The largest palindrome made
// from the product of two 2-digit numbers is 9009 = 91 × 99.
//
// Find the largest palindrome made from the product of two 3-digit numbers.

#include "euler.h"

bool is_palindrome(int num) {
    std::string s = std::to_string(num);
    return std::equal(s.begin(), s.begin() + s.size() / 2, s.rbegin());
}

std::pair<int, int> euler_0004(int limit) {
    int upper = std::pow(10, limit);

    int largest_palindrome{0};
    std::pair<int, int> factors;
    for (int i{100}; i < upper; i++) {
        for (int j{i}; j < upper; j++) {
            int product{i * j};
            if (is_palindrome(product) && (product > largest_palindrome)) {
                largest_palindrome = product;
                factors = std::pair<int, int>(i, j);
            }
        }
    }

    return factors;
}
